src/primitives/hrp.rs

```diff
@@ -119,6 +119,78 @@ impl Hrp {
         Ok(new)
     }
 
+    /// Parses the human-readable part from an object which can be formatted.
+    ///
+    /// The formatted form of the object is subject to all the same rules as [`Self::parse`].
+    /// This method is semantically equivalent to `Hrp::parse(&data.to_string())` but avoids
+    /// allocating an intermediate string.
+    pub fn parse_display<T: core::fmt::Display>(data: T) -> Result<Self, Error> {
+        use Error::*;
+
+        struct ByteFormatter {
+            arr: [u8; MAX_HRP_LEN],
+            index: usize,
+            error: Option<Error>,
+        }
+
+        impl core::fmt::Write for ByteFormatter {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                let mut has_lower: bool = false;
+                let mut has_upper: bool = false;
+                for b in s.bytes() {
+                    // Continue after finding an error so that we report the first invalid
+                    // character, not the last.
+                    if !b.is_ascii() {
+                        self.error = Some(Error::NonAsciiChar(char::from(b)));
+                        continue;
+                    } else if !(33..=126).contains(&b) {
+                        self.error = Some(InvalidAsciiByte(b));
+                        continue;
+                    }
+
+                    if b.is_ascii_lowercase() {
+                        if has_upper {
+                            self.error = Some(MixedCase);
+                            continue;
+                        }
+                        has_lower = true;
+                    } else if b.is_ascii_uppercase() {
+                        if has_lower {
+                            self.error = Some(MixedCase);
+                            continue;
+                        }
+                        has_upper = true;
+                    };
+                }
+
+                // However, an invalid length error will take priority over an
+                // invalid character error.
+                if self.index + s.len() > self.arr.len() {
+                    self.error = Some(Error::TooLong(self.index + s.len()));
+                } else {
+                    // Only do the actual copy if we passed the index check.
+                    self.arr[self.index..self.index + s.len()].copy_from_slice(s.as_bytes());
+                }
+
+                // Unconditionally update self.index so that in the case of a too-long
+                // string, our error return will reflect the full length.
+                self.index += s.len();
+                Ok(())
+            }
+        }
+
+        let mut byte_formatter = ByteFormatter { arr: [0; MAX_HRP_LEN], index: 0, error: None };
+
+        write!(byte_formatter, "{}", data).expect("custom Formatter cannot fail");
```

Comment:
Doesn't this macro call have to allocate a `String` because `data` is not know until runtime?

My understanding is that `write!` calls `format_args!` which creates an [Arguments](https://doc.rust-lang.org/std/fmt/struct.Arguments.html) type which has an `as_str` method but only if there are no runtime args.

In other words isn't this the same performance as:


```rust
    pub fn parse_display<T: core::fmt::Display>(data: T) -> Result<Self, Error> {
        let s = data.to_string();
        Self::parse(&s)
    }
```

Reply:
This code works without any allocator. It's not allocating a string. It would also be insane for the `fmt::Write` API to work this way, where every single call to `write_str` involved a gratuitous allocation.

Reply:
Or indeed, for every call to `write!` anywhere to involve gratuitous allocations.

If you think this could be true we'd need to change our `fmt::Display` implementations throughout every one of our crates to avoid this macro.

Reply:
It took me a little while to trace through the code but the [core of `write!` is the `write` function](https://doc.rust-lang.org/1.80.0/src/core/fmt/mod.rs.html#1149-1197) which iterates over the "pieces" of an `Arguments` and does not allocate.

Reply:
Cool, TIL. Thanks for the link (I was unconvinced before it, you saved me trawling through the source :)